Tools/AP_Periph/battery_bms.cpp

```diff
@@ -0,0 +1,454 @@
+
+    // update LED state machine
+    update_led_state();
+}
+
+// check and handle button press events
+void BatteryBMS::handle_button_press(void)
+{
+    // we take action in these cases
+    //   1. button is released after being pressed between 10ms an 1 sec, we display the battery percentage (short press)
+    //   2. button is pressed for at least 1 second (long press), we start the power-up or power-down counter and animation
+    //   3. button is released after case 2 but before the counter and animation has completed, we cancel the power-up or power-down action
+    //   4. button is released after the power-up or power-down counter and anmation have completed, no further action is taken
+
+    // ignore button presses during startup to avoid false detections from GPIO initialization
+    uint32_t now_ms = AP_HAL::millis();
+    if (!button.startup_complete && (now_ms < BUTTON_STARTUP_DELAY_MS)) {
+        return;
+    }
+    button.startup_complete = true;
+
+    // read current button state (assuming active low - pressed = 0)
+    bool button_pressed = (hal.gpio->read(HAL_GPIO_PIN_BMS_BTN1) == 0);
+
+    // check if button has just been pressed (transition from released to pressed)
+    if (button_pressed && !button.pressed_prev) {
+        printf("BMS: Button pressed!\n");
+
+        // record time button was pressed
+        button.pressed_start_ms = now_ms;
+
+        // record button pressed state before returning
+        button.pressed_prev = button_pressed;
+        button.long_press_handled = false;
+        return;
+    }
+
+    // calculate how long the button was pressed including when button has just been released
+    const uint32_t press_duration_ms = (button_pressed || button.pressed_prev) ? (now_ms - button.pressed_start_ms) : 0;
+    const bool short_press = (press_duration_ms > BUTTON_SHORT_PRESS_THRESHOLD_MS) && (press_duration_ms < BUTTON_LONG_PRESS_THRESHOLD_MS);
+    const bool long_press = (press_duration_ms >= BUTTON_LONG_PRESS_THRESHOLD_MS);
+
+    // check if button has just been released (transition from pressed to released)
+    bool button_released = false;
+    if (!button_pressed && button.pressed_prev) {
+        printf("BMS: Button released!\n");
+        button.pressed_start_ms = 0;
+        button_released = true;
+    }
+
+    // update button state for next iteration
+    button.pressed_prev = button_pressed;
+
+    // handle release after short press
+    // displays battery SOC percentage on LEDs and prints voltages to the console
+    if (short_press && button_released) {
+        request_display_percentage();
+
+        uint8_t batt_soc_pct;
+        if (get_percentage(batt_soc_pct)) {
+            // debug output SOC and cell voltages
+            printf("Battery SOC:%d%% voltage:%.3fV\n", batt_soc_pct, (double)periph.battery_lib.voltage());
+            const AP_BattMonitor::cells &cell_voltages = periph.battery_lib.get_cell_voltages();
+            printf("Cell voltages: ");
```

Comment:
remove debug code, or put behind an #if (not just this line, there are lots more)

Reply:
debug removed in follow-up commit, thanks!